dbms/src/TiDB/Collation/Collator.cpp

```diff
@@ -102,63 +113,270 @@ void Pattern<Collator>::compile(const std::string & pattern, char escape)
-                p_idx++;
-                continue;
+                else if (match_types[p_idx] == Any)
+                {
+                    // Last '%' can match all left characters
+                    if (p_idx == pattern_weights.size() - 1)
+                    {
+                        return true;
+                    }
+                    p_idx_after_any = ++p_idx;
+                    backtrack_s_offset = s_offset;
+                    continue;
+                }
             }
         }
-        if (0 < next_s_offset && next_s_offset <= length)
+        else
         {
-            p_idx = next_p_idx;
-            s_offset = next_s_offset;
+            // All characters in the pattern have been matched,
+            // we need to check if the rest of the string is empty.
+            if (s_offset >= length)
+            {
+                return true;
+            }
+            // Else there are still characters in the string to match,
+            // we need to backtrack below if there is '%' before.
+        }
+
+        // Backtrack
+        if (p_idx_after_any > 0)
+        {
+            Collator::decodeChar(s, s_offset = backtrack_s_offset);
+            // Fast forward to the first match position
+            if (match_types[p_idx_after_any] == Match)
+            {
+                while (true)
+                {
+                    backtrack_s_offset = s_offset;
+                    if (Collator::regexEq(Collator::decodeChar(s, s_offset), pattern_weights[p_idx_after_any]))
+                    {
+                        break;
+                    }
+                    if (s_offset >= length)
+                    {
+                        return false;
+                    }
+                }
+
+                p_idx = p_idx_after_any + 1;
+                continue;
+            }
+            p_idx = p_idx_after_any;
+            backtrack_s_offset = s_offset;
             continue;
         }
         return false;
     }
     return true;
 }
 
+// Similar logical like match, but don't need to decodeChar
+template <typename Collator>
+int Pattern<Collator>::try_match_ascii_ci(const char * s, size_t length) const
+{
+    size_t p_idx = 0;
+    size_t p_idx_after_any = 0;
+    size_t str_idx = 0;
+    size_t backtrack_idx = 0;
+
+    while (true)
+    {
+        if (p_idx < ascii_ci_pattern.size())
+        {
+            if (str_idx >= length)
+            {
+                if (match_types[p_idx] == Any && p_idx == ascii_ci_pattern.size() - 1)
+                {
+                    return 1;
+                }
+                else
+                {
+                    return 0;
+                }
+            }
+            else
+            {
+                if (s[str_idx] < 0)
+                {
+                    return -1;
+                }
+                if ((match_types[p_idx] == Match
+                     && weight_ascii_ci[s[str_idx]] == weight_ascii_ci[ascii_ci_pattern[p_idx]])
```

Comment:
why not just save the converted pattern in `ascii_ci_pattern` during `try_compile_ascii`, then during `match`, there is no need to lookup the `weight_ascii_ci` for pattern chars?